protocol-jackson/README.md

```diff
@@ -0,0 +1,203 @@
+```
+
+To deserialize the record back, you simply need to specify the abstract type, either as a class or
+as a type reference.
+
+```java
+// assume a myJson string containing the JSON
+final ObjectMapper mapper=new ObjectMapper();
+final Record<?> record=mapper.readValue(myJson,AbstractRecord.class);
+
+// record will be correctly deserialized with the right parametric type
+```
+
+### Polymorphic deserialization
+
+Since the main `Record` interface is parameterized over the value property, this means we have to
+figure out concrete type of the `value` object at deserialization time. The protocol defines this
+via the `valueType` property. This is an enum which defines the expected type of the `value` at
+runtime.
+
+To achieve this, the first step is telling Jackson that the concrete type of the `value` property is
+resolved by its sibling property, `valueType`. This is achieved via the `JsonTypeInfo` annotation.
+Then we provide a custom `TypeIdResolver` (`ValueTypeIdResolver` and `IntentTypeIdResolver`) which
+provide Jackson with the concrete type from the given value type during deserialization.
+
+To support this and avoid having to map the value type multiple times, we have a central mapping in
+`ValueTypes`, which maps a `ValueType` constant to a pair of `Intent` sub class and an appropriate
+`RecordValue` sub class.
+
+## Using the builders
+
+As mentioned, one use case is in testing. It can be useful to manually create a series of records as
+input for some test. You can use the generated builders to do so. Builders are named after their
+abstract type, minus the `Abstract` prefix. This means `AbstractRecord` generates `RecordBuilder`,
+`AbstractVariableRecordValue` generates `VariableRecordValueBuilder`, and so on.
+
+The generated builders are plain old builders for beans. To use them, simply instantiate them, fill
+out the attributes, and build the object.
+
+```java
+final VariableDocumentRecordValue value=new VariableDocumentRecordValueBuilder()
+  .scopeKey(1)
+  .semantics(VariableDocumentUpdateSemantic.PROPAGATE)
+  .putVariables(Map.of())
+  .build();
+final RecordBuilder<?> builder=new RecordBuilder<>()
+  .value(value)
+  .valueType(ValueType.VARIABLE_DOCUMENT)
+  .intent(VariableDocumentIntent.UPDATE)
+  .recordType(RecordType.COMMAND)
+  .key(1);
+
+// apply more properties
+final Record<?> record=builder.build();
+```
+
+> NOTE: some attributes are required and do not have a default value. The `#build` method will fail
+> in this case with an appropriate error. This isn't fixed in stone, and if we want we can always
+> add more sane defaults, but it was decided to go with this for now to simplify the initial
+> iteration.
+
+Also note that builders have a `#from` method generated which accepts all super types of the
+abstract type. So for `VariableRecordValueBuilder`, there would be a `#from(VariableRecordValue)`
+method generated which can accept any implementations of the interface.
```

Comment:
💭 Could be nice to add here that this is an alternative to copy and provides complete deep copying. However, given the immutable nature of the generated classes, copy can also be used directly for faster copying.

Reply:
It doesn't provide deep copying actually. If you got that from this, then I need to rewrite it :smile: